refactor template order `allocMappedBuf`

SimeonEhrig opened the pull request:
Move template TPlatform as the last template. There is no need to provide the platform template signature if we pass the platform as an instance.

follow up of #2162

psychocoderHPC commented:
please update https://github.com/alpaka-group/alpaka/blob/8dbe7c42c647e4b67f22c10a39c36b160f654ecf/docs/source/dev/backends.rst?plain=1#L386 too.
It was also forgotten during the change from platform type to platform instance.

Maybe we should use here `allocMappedBufIfSupported()`

psychocoderHPC commented:
please update https://github.com/alpaka-group/alpaka/blob/8dbe7c42c647e4b67f22c10a39c36b160f654ecf/docs/source/dev/backends.rst?plain=1#L366 too


```c++
alpaka::allocMappedBuf<TElement>(host, platform, extents) 1D, 2D, 3D supported!  
```

SimeonEhrig commented:
> please update
> 
> https://github.com/alpaka-group/alpaka/blob/8dbe7c42c647e4b67f22c10a39c36b160f654ecf/docs/source/dev/backends.rst?plain=1#L366
> too
> 
> ```c++
> alpaka::allocMappedBuf<TElement>(host, platform, extents) 1D, 2D, 3D supported!  
> ```

I `alpaka::allocMappedBuf<TElement, TIdx>(host, platform, extents) 1D, 2D, 3D supported!` should be correct.

psychocoderHPC commented on docs/source/dev/backends.rst:
```diff
@@ -363,7 +363,7 @@ The following tables list the functions available in the `CUDA Runtime API <http
     | cudaGetSymbolSize          | --                                                                                         |
     +----------------------------+--------------------------------------------------------------------------------------------+
-    | cudaHostAlloc              | alpaka::allocMappedBuf<TPlatform, TElement>(host, extents) 1D, 2D, 3D supported!           |
+    | cudaHostAlloc              | alpaka::allocMappedBuf<TElement, TIdx>(host, extents) 1D, 2D, 3D supported!                |
```
Sry my comment was wrong this should be `allocMappedBufIfSupported`. There is no strong need that this memory is mapped.
`allocMappedBuf` is not available for any accelerator or device.

fwyzard commented on docs/source/dev/backends.rst:
```diff
@@ -363,7 +363,7 @@ The following tables list the functions available in the `CUDA Runtime API <http
     | cudaGetSymbolSize          | --                                                                                         |
     +----------------------------+--------------------------------------------------------------------------------------------+
-    | cudaHostAlloc              | alpaka::allocMappedBuf<TPlatform, TElement>(host, extents) 1D, 2D, 3D supported!           |
+    | cudaHostAlloc              | alpaka::allocMappedBufIfSupported<TElement, TIdx>(host, extents) 1D, 2D, 3D supported!     |
```
This too should be just
```suggestion
    | cudaHostAlloc              | alpaka::allocMappedBuf<TElement, TIdx>(host, extents) 1D, 2D, 3D supported!                |
```

psychocoderHPC replied:
I do not think that `allocMappedBuf` is correct here. `allocMappedBuf` can fail if mapped memory is not supported for the current device, so the code will not run with all devices. `allocMappedBufIfSupported` will be safe for all kinds of devices.

The CUDA API function `cudaHostAlloc` is pinning memory but does not guarantee that the memory is visible on the GPU device, so the best fitting alpaka equivalent is `allocMappedBufIfSupported`

fwyzard replied:
> The CUDA API function `cudaHostAlloc` is pinning memory but does not guarantee that the memory is visible on the GPU device, so the best fitting alpaka equivalent is `allocMappedBufIfSupported`

It does guarantee that. From the [CUDA Runtime API](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1gb65da58f444e7230d3322b6126bb4902):

> `__host__ cudaError_t cudaHostAlloc ( void** pHost, size_t size, unsigned int  flags )`
>
> #### Description
>
> Allocates size bytes of host memory that is page-locked and accessible to the device.

fwyzard replied:
In any case, [`cudaMallocHost()`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__HIGHLEVEL.html#group__CUDART__HIGHLEVEL_1gd5c991beb38e2b8419f50285707ae87e) and [`cudaHostAlloc()`](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1gb65da58f444e7230d3322b6126bb4902) are identical, so they should map to the same functionality in alpaka.

psychocoderHPC replied:
Ohh I was not aware that `cudaMallocHost` is the C++-API equivalent of `cudaHostAlloc`. I used always `cudaHostAlloc` with flags to create mapped memory.

In this case, IMO both should be equal. The problem I had is that for both CUDA methods, you can configure the behavior via a flag where we have only named functions. Maybe we should add  `allocMappedBufIfSupported`, ` allocMappedBuf`, and `allocMappedBuf` into the documentation and the user must decide what fitts best for their problem because the method used for the replacement depends on the flags used in the native CUDA function call.

SimeonEhrig replied:
I changed it back to `alpaka::allocMappedBuf<TPlatform, TElement>(host, extents)` and wrote a for foot note for `alpaka::allocMappedBufIfSupported<TPlatform, TElement>(host, extents)`

fwyzard commented on docs/source/dev/backends.rst:
```diff
@@ -363,7 +363,7 @@ The following tables list the functions available in the `CUDA Runtime API <http
     | cudaGetSymbolSize          | --                                                                                         |
     +----------------------------+--------------------------------------------------------------------------------------------+
-    | cudaHostAlloc              | alpaka::allocMappedBuf<TPlatform, TElement>(host, extents) 1D, 2D, 3D supported!           |
+    | cudaHostAlloc              | alpaka::allocMappedBuf<TElement, TIdx>(host, extents) 1D, 2D, 3D supported! [1]            |
```
```suggestion
    | cudaHostAlloc              | alpaka::allocMappedBuf<TElement, TIdx>(host, platform, extents) 1D, 2D, 3D supported! [1]  |
```

fwyzard replied:
also below

SimeonEhrig committed:
refactor template order `allocMappedBuf`

Move template TPlatform as the last template. There is no need to provide the platform template signature if we pass the platform as an instance.
